### PR TITLE
[SPARK-33979][SQL] Reorder predicate

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This pr add support reorder predicates based on selectivity and compute cost.

1. Conjunctive predicates
   Formula: `(selectivity - 1.0) / expression cost`
2. Disjunctive predicates
   Formula: `(-selectivity) / expression cost`

 
### Why are the changes needed?

Improve filter performance. Many mainstream databases support this optimization:

https://issues.apache.org/jira/browse/HIVE-21857
https://issues.apache.org/jira/browse/IMPALA-2805

https://www.ibm.com/support/knowledgecenter/SSSHTQ_8.1.0/com.ibm.netcool_OMNIbus.doc_8.1.0/omnibus/wip/admin/reference/omn_adm_per_optimizationrules.html#omn_adm_per_optimizationrules__reorder
https://docs.oracle.com/en/database/oracle/oracle-database/21/addci/extensible-optimizer-interface.html#GUID-28A4EDA6-19DD-4773-B3B8-1802C3B01E21
 https://docs.oracle.com/cd/B10501_01/server.920/a96533/hintsref.htm#13676

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test and Benchmark test.

Benchmark by `Benchmark` framework.
```scala
val numRows = 1024 * 1024 * 15
val mid = numRows / 2
spark.range(numRows).selectExpr("id as a", "id as b", "id as c").write.saveAsTable("t1")
spark.sql("ANALYZE TABLE t1 COMPUTE STATISTICS FOR ALL COLUMNS")
val benchmark1 = new org.apache.spark.benchmark.Benchmark("Benchmark predicate reorder(Reduce MultiLikeBase priority)", numRows, minNumIters = 5)
Seq(false, true).foreach { pushDownEnabled =>
  Seq(false, true).foreach { reorderEnabled =>
    val name = s"Parquet Vectorized ${if (pushDownEnabled) s"(pushDownEnabled)" else ""} ${if (reorderEnabled) s"(reorderEnabled)" else ""}"
    benchmark1.addCase(name) { _ =>
      withSQLConf(
        SQLConf.CBO_ENABLED.key -> "true",
        SQLConf.PARQUET_FILTER_PUSHDOWN_ENABLED.key -> s"$pushDownEnabled",
        SQLConf.PREDICATE_REORDER_ENABLED.key -> s"$reorderEnabled") {
        spark.sql(s"SELECT * FROM t1 WHERE cast(a as string) like all (${Range(1, 500).map(s => s"'%${s}%test%'").mkString(", ")}) and c > $mid and b < ${mid + 10}").write.format("noop").mode("Overwrite").save()
      }
    }
  }
}
benchmark1.run()

val benchmark2 = new org.apache.spark.benchmark.Benchmark("Benchmark predicate reorder(Reorder predicates by selectivity)", numRows, minNumIters = 5)
Seq(false, true).foreach { pushDownEnabled =>
  Seq(false, true).foreach { reorderEnabled =>
    val name = s"Parquet Vectorized ${if (pushDownEnabled) s"(pushDownEnabled)" else ""} ${if (reorderEnabled) s"(reorderEnabled)" else ""}"
    benchmark2.addCase(name) { _ =>
      withSQLConf(
        SQLConf.CBO_ENABLED.key -> "true",
        SQLConf.PARQUET_FILTER_PUSHDOWN_ENABLED.key -> s"$pushDownEnabled",
        SQLConf.PREDICATE_REORDER_ENABLED.key -> s"$reorderEnabled") {
        spark.sql(s"SELECT * FROM t1 WHERE a > 10 and b > 20000 and c > $mid and b < ${mid + 10}").write.format("noop").mode("Overwrite").save()
      }
    }
  }
}
benchmark2.run()
```
```
Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.15.7
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Benchmark predicate reorder(Reduce MultiLikeBase priority):  Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------------------------
Parquet Vectorized                                                   3872           4100         249          4.1         246.2       1.0X
Parquet Vectorized  (reorderEnabled)                                  453            479          23         34.7          28.8       8.6X
Parquet Vectorized (pushDownEnabled)                                 4045           4707        1010          3.9         257.2       1.0X
Parquet Vectorized (pushDownEnabled) (reorderEnabled)                 426            489          42         37.0          27.1       9.1X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.15.7
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Benchmark predicate reorder(Reorder predicates by selectivity):  Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
----------------------------------------------------------------------------------------------------------------------------------------------
Parquet Vectorized                                                        541            664         148         29.1          34.4       1.0X
Parquet Vectorized  (reorderEnabled)                                      484            537          52         32.5          30.8       1.1X
Parquet Vectorized (pushDownEnabled)                                      409            479          71         38.5          26.0       1.3X
Parquet Vectorized (pushDownEnabled) (reorderEnabled)                     368            396          20         42.7          23.4       1.5X
```

### CODE REVIEW

File: sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q13.sf100/explain.txt [hunk 1]
```diff
@@ -110,16 +110,16 @@ Input [13]: [ss_sold_date_sk#1, ss_hdemo_sk#3, ss_addr_sk#4, ss_store_sk#5, ss_q
 (16) Scan parquet default.customer_address
 Output [3]: [ca_address_sk#18, ca_state#19, ca_country#20]
 Batched: true
-Location [not included in comparison]/{warehouse_dir}/date_dim]
-PushedFilters: [EqualTo(d_year,2001), IsNotNull(d_year), IsNotNull(d_date_sk)]
-ReadSchema: struct<d_date_sk:int,d_year:int>
+Location [not included in comparison]/{warehouse_dir}/customer_address]
+PushedFilters: [Or(Or(In(ca_state, [TX,OH]),In(ca_state, [OR,NM,KY])),In(ca_state, [VA,TX,MS])), IsNotNull(ca_country), EqualTo(ca_country,United States), IsNotNull(ca_address_sk)]
+ReadSchema: struct<ca_address_sk:int,ca_state:string,ca_country:string>
```
Comment (wangyum) [inline sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q13.sf100/explain.txt]: @Ngone51 This is not the expected change, there is a bug in PlanStability?

Comment (Ngone51) [reply sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q13.sf100/explain.txt]: Is it still an issue? I see it's marked as `Outdated`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+          .map(_._1).reduceLeft(And)
+    }
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    if (!conf.predicateReorder) {
+      plan
+    } else {
+      plan transform {
+        case f @ Filter(cond, _) =>
```
Comment (tanelk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala:59]: Should it be done also in the Join condition?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 3]
```diff
@@ -199,6 +199,8 @@ abstract class Optimizer(catalogManager: CatalogManager)
     // idempotence enforcement on this batch. We thus make it FixedPoint(1) instead of Once.
     Batch("Join Reorder", FixedPoint(1),
       CostBasedJoinReorder) :+
+    Batch("Reorder Predicate", FixedPoint(1),
```
Comment (tanelk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:202]: Does `Once` fail any tests? It's best to make batches idempotent - it's a helpful test feature.

File: sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q13.sf100/explain.txt (hunk 1 above)
Comment (wangyum) [reply sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q13.sf100/explain.txt]: Thank you, it seems this is not an issue.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 3 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Thank you @tanelk, change it to Once.

**Comment by tanelk** [comment]
It would be nice to do this recursively.
Currently you split the outer `and` filter and sort them. each of these can have an `or` filter and it would be beneficial to sort its children also. Inside `or` the filter selectivity rule must be the other way around.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * A rule that reorder predicate expressions to get better performance.
+ */
+object PredicateReorder extends Rule[LogicalPlan] with PredicateHelper {
+  // Get priority from a Expression. Expressions with higher priority are executed in preference
+  // to expressions with lower priority.
+  private def getPriority(exp: Expression, filterEstimation: Option[FilterEstimation]) = exp match {
+    case e: Expression if e.find(_.isInstanceOf[DynamicPruningSubquery]).isDefined => 1.0
```
Comment (tanelk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala:32]: What about other instances of `SubqueryExpression`? they are also likely to be very slow compared to simple expressions.

**Comment by tanelk** [comment]
> It would be nice to do this recursively.
> Currently you split the outer `and` filters and sort them. each of these can have an `or` filter as a child and it would be beneficial to split and sort those too. Inside `or` the filter selectivity rule must be the other way around.

I was thinking something like this.

```diff
diff --git a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala
index 805f9afeba..345c30896f 100644
--- a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala
+++ b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala
@@ -28,7 +28,10 @@ import org.apache.spark.sql.catalyst.rules.Rule
 object PredicateReorder extends Rule[LogicalPlan] with PredicateHelper {
   // Get priority from a Expression. Expressions with higher priority are executed in preference
   // to expressions with lower priority.
-  private def getPriority(exp: Expression, filterEstimation: Option[FilterEstimation]) = exp match {
+  private def getPriority(
+      exp: Expression,
+      filterEstimation: Option[FilterEstimation],
+      conjuction: Boolean) = exp match {
     case e: Expression if e.find(_.isInstanceOf[SubqueryExpression]).isDefined => 1.0
     case e: Expression if e.find(_.isInstanceOf[UserDefinedExpression]).isDefined => 2.0
     case e: Expression if e.find(_.isInstanceOf[MultiLikeBase]).isDefined ||
@@ -38,8 +41,11 @@ object PredicateReorder extends Rule[LogicalPlan] with PredicateHelper {
         case c: CaseWhen => c.branches.size
       }.max
       3.0 + (1.0 / maxSize)
-    case e =>
+    case e if conjuction =>
       filterEstimation.flatMap(_.calculateFilterSelectivity(e, false).map(5.0 - _)).getOrElse(4.0)
+    case e if !conjuction =>
+      filterEstimation.flatMap(_.calculateFilterSelectivity(e, false).map(4.0 + _)).getOrElse(4.0)
+
   }

   private def reorderPredicates(e: Expression, estimation: Option[FilterEstimation]): Expression = {
@@ -47,10 +53,16 @@ object PredicateReorder extends Rule[LogicalPlan] with PredicateHelper {
       case _: Or =>
         splitDisjunctivePredicates(e)
           .map(reorderPredicates(_, estimation))
+          .map(e => (e, getPriority(e, estimation, conjuction = false)))
+          .sortWith(_._2 > _._2)
+          .map(_._1)
           .reduceLeft(Or)
       case _: And =>
         splitConjunctivePredicates(e)
-          .map(e => (e, getPriority(e, estimation))).sortWith(_._2 > _._2).map(_._1)
+          .map(reorderPredicates(_, estimation))
+          .map(e => (e, getPriority(e, estimation, conjuction = true)))
+          .sortWith(_._2 > _._2)
+          .map(_._1)
           .reduceLeft(And)
       case _ => e
     }
```

**Comment by wangyum** [comment]
```scala
spark.range(20000).selectExpr("id as a", "id as b", "id as c").write.saveAsTable("t1")
spark.sql("ANALYZE TABLE t1 COMPUTE STATISTICS FOR ALL COLUMNS")
spark.sql(s"SELECT * FROM t1 WHERE (a > 10 and b > 10000) or (c < 10 and b < 10000)").explain("cost")
```
My change result:
```
== Optimized Logical Plan ==
Filter (((b#3193L > 10000) AND (a#3192L > 10)) OR ((c#3194L < 10) AND (b#3193L < 10000))), Statistics(sizeInBytes=312.4 KiB, rowCount=1.00E+4)
+- Relation[a#3192L,b#3193L,c#3194L] parquet, Statistics(sizeInBytes=625.0 KiB, rowCount=2.00E+4)
```
@tanelk's  change result:
```
== Optimized Logical Plan ==
Filter (((a#3192L > 10) AND (b#3193L > 10000)) OR ((c#3194L < 10) AND (b#3193L < 10000))), Statistics(sizeInBytes=312.4 KiB, rowCount=1.00E+4)
+- Relation[a#3192L,b#3193L,c#3194L] parquet, Statistics(sizeInBytes=625.0 KiB, rowCount=2.00E+4)
```

It seems my change is better.

**Comment by wangyum** [comment]
Another related report: [Adaptive filter ordering in Spark](https://arxiv.org/pdf/1905.01349.pdf).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala (hunk 4 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala]: yes, add it.

**Comment by tanelk** [comment]
Thats strange, I ran the same test and got this optimized plan for my example:

```
== Optimized Logical Plan ==
Filter (((b#3653L > 10000) AND (a#3652L > 10)) OR ((c#3654L < 10) AND (b#3653L < 10000))), Statistics(sizeInBytes=312.4 KiB, rowCount=1.00E+4)
+- Relation[a#3652L,b#3653L,c#3654L] parquet, Statistics(sizeInBytes=625.0 KiB, rowCount=2.00E+4)
````

Something the test does not illustrate is that my proposition can swap the sides of the OR.

> ```scala
> spark.range(20000).selectExpr("id as a", "id as b", "id as c").write.saveAsTable("t1")
> spark.sql("ANALYZE TABLE t1 COMPUTE STATISTICS FOR ALL COLUMNS")
> spark.sql("set spark.sql.cbo.enabled=true")
> spark.sql(s"SELECT * FROM t1 WHERE (a > 10 and b > 10000) or (c < 10 and b < 10000)").explain("cost")
> ```
> 
> My change result:
> 
> ```
> == Optimized Logical Plan ==
> Filter (((b#3193L > 10000) AND (a#3192L > 10)) OR ((c#3194L < 10) AND (b#3193L < 10000))), Statistics(sizeInBytes=312.4 KiB, rowCount=1.00E+4)
> +- Relation[a#3192L,b#3193L,c#3194L] parquet, Statistics(sizeInBytes=625.0 KiB, rowCount=2.00E+4)
> ```
> 
> @tanelk's change result:
> 
> ```
> == Optimized Logical Plan ==
> Filter (((a#3192L > 10) AND (b#3193L > 10000)) OR ((c#3194L < 10) AND (b#3193L < 10000))), Statistics(sizeInBytes=312.4 KiB, rowCount=1.00E+4)
> +- Relation[a#3192L,b#3193L,c#3194L] parquet, Statistics(sizeInBytes=625.0 KiB, rowCount=2.00E+4)
> ```
> 
> It seems my change is better.

**Comment by cloud-fan** [comment]
> Move UDF, LikeAny, LikeAll and CaseWhen to the end because these Expressions always CPU intensive.

I'm not sure about it. CaseWhen can be fast if it has only one branch. Probably we should only do this for CBO.

**Comment by wangyum** [comment]
Ok, make it only work in CBO mode.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.plans.logical.statsEstimation.FilterEstimation
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * A rule that reorder predicate expressions to get better performance.
+ */
+object PredicateReorder extends Rule[LogicalPlan] with PredicateHelper {
+
+  // Return 1.0D - 1.0D / Int.MaxValue if the expression is not supported calculate
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala:30]: `1.0D - 1.0D / Int.MaxValue` is NaN or Infinity? Can we use Double constant?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    (selectivity(exp, filterEstimation) - 1.0D) / expressionCost(exp)
+  }
+
+  // Formula: (-selectivity) / expression cost
+  private def rankingOr(exp: Expression, filterEstimation: FilterEstimation): Double = {
+    -selectivity(exp, filterEstimation) / expressionCost(exp)
+  }
+
+  // The cost of a call expression e is computed as:
+  //   cost(exp) = typeSize + functionCost + cost(children).
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala:48]: how do we define `functionCost`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+          .map(e => (e, rankingAnd(e, filterEstimation))).sortWith(_._2 < _._2).map(_._1)
+          .reduceLeft(And)
+      case _ => exp
+    }
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    if (conf.cboEnabled && conf.predicateReorder) {
+      plan transform {
+        case f @ Filter(cond, _) if cond.references.toSet.subsetOf(f.stats.attributeStats.keySet) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala:94]: shall we do the same thing for join condition? And conditions in `If`, `CaseWhen`?

**Comment by cloud-fan** [comment]
Since this changes the TPCDS plan golden file, do we have some TPCDS benchmark results?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala (hunk 5 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala:30]: `1.0D - 1.0D / Int.MaxValue` is 0.9999999995343387:
```
scala> 1.0D - 1.0D / Int.MaxValue
res25: Double = 0.9999999995343387

scala> 1.0D - 1.0D / Double.MaxValue
res26: Double = 1.0

scala> 1.0D - 1.0D / Long.MaxValue
res27: Double = 1.0
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala (hunk 6 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala:67]: The `functionCost` refers to Hive and Impala:
https://github.com/apache/hive/blob/91a92e323a29a49b5ad3b860b9523f347b6b8d4c/ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterSortPredicates.java#L243-L281
https://github.com/apache/impala/blob/d271baa33da1a02aa6ffc47b0380dc62239107b4/fe/src/main/java/org/apache/impala/analysis/Expr.java#L79-L94

**Comment by wangyum** [comment]
> Since this changes the TPCDS plan golden file, do we have some TPCDS benchmark results?

In fact, it will not have significant improvement in TPC-DS query, because the filter conditions of the TPC-DS query are very simple.

---

This is a real query from cluster. The order of filter conditions for this query is:
1. LIKE ANY with 70 values
2. In predicate with 702 values
3. CAST(D.AUCT_END_DATE as DATE)>=current_date-65

We will reorder the filter conditions with this optimizer:
1. CAST(D.AUCT_END_DATE as DATE)>=current_date-65
2. In predicate with 702 values
3. LIKE ANY with 70 values



Before this pr | After this pr
-- | --
13 min | More than 72 min



image

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala (hunk 7 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala]: Could we do it step by step?

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala]: Could we do it step by step?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+ * A rule that reorder conditions in filters base on estimated selectivity and compute cost.
+ *
+ * {{{
+ *   SELECT * FROM lineitem WHERE l_comment LIKE '%a%' AND l_orderkey = 1024 ==>
+ *   SELECT * FROM lineitem WHERE l_orderkey = 1024 AND l_comment LIKE '%a%'
+ * }}}
+ */
+object PredicateReorder extends Rule[LogicalPlan] with PredicateHelper {
+
+  private val DEFAULT_SELECTIVITY = 0.1D
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala:35]: This default value from Impala: https://github.com/apache/impala/blob/d271baa33da1a02aa6ffc47b0380dc62239107b4/fe/src/main/java/org/apache/impala/analysis/Expr.java#L76-L77

Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala:35]: This default value from Impala: https://github.com/apache/impala/blob/d271baa33da1a02aa6ffc47b0380dc62239107b4/fe/src/main/java/org/apache/impala/analysis/Expr.java#L76-L77

Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala:35]: This default value from Impala: https://github.com/apache/impala/blob/d271baa33da1a02aa6ffc47b0380dc62239107b4/fe/src/main/java/org/apache/impala/analysis/Expr.java#L76-L77

Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala:35]: This default value from Impala: https://github.com/apache/impala/blob/d271baa33da1a02aa6ffc47b0380dc62239107b4/fe/src/main/java/org/apache/impala/analysis/Expr.java#L76-L77

Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala:35]: This default value from Impala: https://github.com/apache/impala/blob/d271baa33da1a02aa6ffc47b0380dc62239107b4/fe/src/main/java/org/apache/impala/analysis/Expr.java#L76-L77

Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala:35]: This default value from Impala: https://github.com/apache/impala/blob/d271baa33da1a02aa6ffc47b0380dc62239107b4/fe/src/main/java/org/apache/impala/analysis/Expr.java#L76-L77

May be we can improve the default selectivity estimate in the future:
https://github.com/postgres/postgres/blob/ed934d4fa30f0f94e6f7125ad2154e6a58d1c7f7/src/include/utils/selfuncs.h#L33-L49

**Comment by karenfeng** [comment]
@wangyum, what is the status of this work? Would be nice to have this feature.

**Comment by wangyum** [comment]
@karenfeng Our production environment is already using this feature.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    -filterSelectivity(exp, filterEstimation) / expressionCost(exp)
+  }
+
+  // The cost of a call expression exp is computed as:
+  //   cost(exp) = typeSize + functionCost + cost(children).
+  private def expressionCost(exp: Expression): Double = exp match {
+      case e: Expression if e.children.isEmpty =>
+        e.dataType.defaultSize
+      case e: IsNull =>
+        e.dataType.defaultSize + IS_OP_COST + e.children.map(expressionCost).sum
```
Comment (karenfeng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala:72]: If I'm reading this right, `e.dataType.defaultSize` will (usually) have a heavier weight than the operator cost. This [matches](https://github.com/apache/hive/blob/7b3ecf617a6d46f48a3b6f77e0339fd4ad95a420/ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterSortPredicates.java#L237) Hive, but [not](https://github.com/apache/impala/blob/b28da054f3595bb92873433211438306fc22fbc7/fe/src/main/java/org/apache/impala/analysis/IsNullPredicate.java#L165) Impala. Have you seen any issues with this?

**Review by karenfeng — COMMENTED** [review]
Thanks for the update @wangyum! From what I can tell, the expression cost calculation here matches Hive's implementation (by always accounting for data type size), while the operation cost values are about halfway in between those from Impala/Hive. Did you find that Impala underestimated cost due to data type size, or that Hive's operation cost values were overestimated?

**Pull request closed without merging**